mongodbatlas/resource_mongodbatlas_cluster.go

```diff
@@ -1361,22 +1376,188 @@ func flattenReplicationSpecs(rSpecs []matlas.ReplicationSpec) []map[string]inter
+	var providerName, backingProviderName string
+
+	if _, ok := d.GetOk("replication_specs"); !ok {
+		return actualSpecs
+	}
+
+	if val, ok := d.GetOk("provider_name"); ok {
+		providerName = val.(string)
+	}
+
+	if val, ok := d.GetOk("backing_provider_name"); ok {
+		backingProviderName = val.(string)
+	}
+
+	stateReplicationSpecsMap := getStateFileSpecsMap(d.Get("replication_specs"), providerName, backingProviderName)
+
+	for _, actualSpec := range actualSpecs {
+		hash := getReplicationSpecHash(actualSpec, true)
+		matchedStateReplicationSpec := stateReplicationSpecsMap[hash]
+
+		for _, actualRegionConfigElem := range actualSpec["regions_config"].([]map[string]interface{}) {
+			actualRegionConfig := actualRegionConfigElem
+			regionConfigHash := getRegionConfigHash(actualRegionConfig)
+			matchedStateRegionConfig := matchedStateReplicationSpec[regionConfigHash].(map[string]interface{})
+
+			actualRegionConfig["region_name"] = matchedStateRegionConfig["region_name_by_user"]
+		}
+	}
+
+	return actualSpecs
+}
+
+// This method creates a map of all the replication_specs present in Terraform state for the cluster
+// keyed on a hash value generated from property values of a replication_spec(s).
+//
+// The value of each entry is another map keyed on a hash value generated from property values of each region_config(s) in the replication_spec
+//
+// A "region_name_by_user" property is also added to each region_config to keep track if the user is
+// using an Atlas region name or a Cloud Provider's region name.
+func getStateFileSpecsMap(stateSpecs interface{}, providerName, backingProviderName string) map[uint64]map[uint64]interface{} {
+	var stateSpec map[string]interface{}
+	var stateRegionsConfigs []interface{}
+
+	stateReplicationSpecsMap := make(map[uint64]map[uint64]interface{})
+
+	for _, s := range stateSpecs.(*schema.Set).List() {
+		stateSpec = s.(map[string]interface{})
+		stateRegionsConfigs = stateSpec["regions_config"].(*schema.Set).List()
+		stateRegionConfigsMap := make(map[uint64]interface{})
+
+		for _, r := range stateRegionsConfigs {
+			stateRegionConfig := r.(map[string]interface{})
+			// we change region_name to Atlas region so we can use this to match with actual regionConfigs received from the API
+			stateRegionConfig["region_name_by_user"] = stateRegionConfig["region_name"]
+			stateRegionConfig["region_name"], _ = GetAtlasRegion(providerName, backingProviderName, stateRegionConfig["region_name_by_user"].(string))
+
+			regionConfigHashKey := getRegionConfigHash(stateRegionConfig)
+			stateRegionConfigsMap[regionConfigHashKey] = stateRegionConfig
+		}
+		repSpecsHashKey := getReplicationSpecHash(stateSpec, true)
+		stateReplicationSpecsMap[repSpecsHashKey] = stateRegionConfigsMap
+	}
+
+	return stateReplicationSpecsMap
+}
+
+// excluding 'id' property when generating hash of a ReplicationSpec
+func getReplicationSpecHash(repSpec map[string]interface{}, stateSpec bool) uint64 {
```

Comment:
cluster is not the only resource that uses provider regions. Should we update other resources as well to use these functions?

Reply:
I would address other resources separately. For this ticket I think we should focus on only mongodbatlas_cluster resource, given the complexity in handling replication specs/region configs. 

We can look into other resources such as mongodbatlas_advanced_cluster and others after this.

cc @Zuhairahmed

Reply:
while agree @maastha that we can we have separate tickets for future unrelated resources, given how we are actively directing users to leverage advanced_clusters resource this PR should address support in both cluster and advanced_cluster